## CS 559 Homework 4

### Nan Chen

### SID: 20018997

### Email: nchen15@stevens.edu

In [1]:
import numpy as np
import pandas as pd

1.	(5 pts) Determine the data type of features.

In [2]:
df = pd.read_csv('Titanic_train.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


2.	(5 pts) Determine the ratio of survival rate. The ideal binary classification situation (e.g., T/F or 1/0) has a 50-to-50 ratio, which we call the balanced set. 

In [3]:
df['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

3.	(5 pts) Determine the missing value features and their number of observations. Explain how you would handle features. Explain whether deleting the feature(s) or observation(s) is the best choice. If not, explain why not. 

In [4]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Age: Since there are 890 entries, and only 177 entry whose age is missing, I will use average age to fill in the NA value. Also, age should be an important measurement of whethere they can survive.

Cabin: Since the majority of the entry is missing the value, I will delete the feature.

Embarked: Only 2 are missing, and it's an object dType value, so I will choose the value with most appearance for this the missing values.

4.	(5 pts) Determine any feature(s) that can be removed. Explain why. Then remove them.

`Name`, `Ticket`, `Fare`, and `Cabin` can be removed, because they are either for indexing which doesn't help with classification, or the price for the ticket that is duplicated with `Pclass`.

In [5]:
df1 = df.drop(['Name','Ticket', 'Fare', "Cabin"], axis=1)
df1.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch Embarked
0            1         0       3    male  22.0      1      0        S
1            2         1       1  female  38.0      1      0        C
2            3         1       3  female  26.0      0      0        S
3            4         1       1  female  35.0      1      0        S
4            5         0       3    male  35.0      0      0        S

5.	(5 pts) Convert text feature(s) into a numerical feature(s). Explain if any features do not require numerical conversion. 

In [6]:
df1['Sex'] = df1['Sex'].map({'male': 0, 'female': 1})
df1['Embarked'] = df1['Embarked'].map({'C': 0, 'Q': 1, 'S': 2})
df1.head()

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch  Embarked
0            1         0       3    0  22.0      1      0       2.0
1            2         1       1    1  38.0      1      0       0.0
2            3         1       3    1  26.0      0      0       2.0
3            4         1       1    1  35.0      1      0       2.0
4            5         0       3    0  35.0      0      0       2.0

I don't think any features do not requre numerical conversation.

6.	(10 pts) Write a method “data_split(data, ratio)” that shuffles the observations and splits the input data set by the input ratio between 0 and 1. Let the method returns the split data sets. 

In [7]:
def data_split(data, ratio):
    shuffled_indices = np.random.permutation(len(data))
    data = data.iloc[shuffled_indices]

    split_index = int(len(data) * ratio)
    train_data = data.iloc[:split_index]
    test_data = data.iloc[split_index:]
    
    return train_data, test_data

7.	(10 pts) Write a method “k_fold_CV(data, k)” that shuffles the observations and returns k many validation sets. But the repetition of observations must be avoided. 

In [8]:
def k_fold_CV(data, k):
    shuffled_indices = np.random.permutation(len(data))
    fold_size = len(data) // k
    validation_sets = []

    for i in range(k):
        start = i * fold_size
        if (i < k - 1):
            end = (i + 1) * fold_size
        else:
            end = len(data)
        
        val_indices = shuffled_indices[start:end]
        val_data = data.iloc[val_indices]
        validation_sets.append(val_data)
    
    return validation_sets